x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/AbstractMultivalueFunctionTestCase.java

```diff
@@ -7,129 +7,446 @@
+    /**
+     * Build a test case with {@code boolean} values.
+     */
+    protected static void booleans(
+        List<TestCaseSupplier> cases,
+        String name,
+        String evaluatorName,
+        DataType expectedDataType,
+        BiFunction<Integer, Stream<Boolean>, Matcher<Object>> matcher
+    ) {
+        cases.add(
+            new TestCaseSupplier(
+                name + "(false)",
+                () -> new TestCase(
+                    List.of(new TypedData(List.of(false), DataTypes.BOOLEAN, "field")),
+                    evaluatorName + "[field=Attribute[channel=0]]",
+                    expectedDataType,
+                    matcher.apply(1, Stream.of(false))
+                )
+            )
+        );
+        cases.add(
+            new TestCaseSupplier(
+                name + "(true)",
+                () -> new TestCase(
+                    List.of(new TypedData(List.of(true), DataTypes.BOOLEAN, "field")),
+                    evaluatorName + "[field=Attribute[channel=0]]",
+                    expectedDataType,
+                    matcher.apply(1, Stream.of(true))
+                )
+            )
+        );
+        for (Block.MvOrdering ordering : Block.MvOrdering.values()) {
+            cases.add(new TestCaseSupplier(name + "(<booleans>) " + ordering, () -> {
+                List<Boolean> mvData = randomList(2, 100, ESTestCase::randomBoolean);
+                putInOrder(mvData, ordering);
+                return new TestCase(
+                    List.of(new TypedData(mvData, DataTypes.BOOLEAN, "field")),
+                    evaluatorName + "[field=Attribute[channel=0]]",
+                    expectedDataType,
+                    matcher.apply(mvData.size(), mvData.stream())
+                );
+            }));
+        }
+    }
 
     /**
-     * Matcher for single valued fields.
+     * Build a test case with {@link BytesRef} values.
      */
-    private Matcher<Object> singleValueMatcher(Object o, DataType dataType) {
-        return o == null ? nullValue() : resultMatcherForInput(List.of(o), dataType);
+    protected static void bytesRefs(
+        List<TestCaseSupplier> cases,
+        String name,
+        String evaluatorName,
+        BiFunction<Integer, Stream<BytesRef>, Matcher<Object>> matcher
+    ) {
+        bytesRefs(cases, name, evaluatorName, DataTypes.KEYWORD, matcher);
     }
 
-    @Override
-    protected final List<AbstractScalarFunctionTestCase.ArgumentSpec> argSpec() {
-        return List.of(required(supportedTypes()));
+    /**
+     * Build a test case with {@link BytesRef} values.
+     */
+    protected static void bytesRefs(
+        List<TestCaseSupplier> cases,
+        String name,
+        String evaluatorName,
+        DataType expectedDataType,
+        BiFunction<Integer, Stream<BytesRef>, Matcher<Object>> matcher
+    ) {
+        cases.add(
+            new TestCaseSupplier(
+                name + "(\"\")",
```

Comment:
On the original PR @not-napoleon commented that this'd be better written as `(empty string)`. Without the context the `""` this renders as is not great.

Reply:
I'll grab this in a followup too.